src/stan/model/indexing/lvalue.hpp

```diff
@@ -118,62 +118,58 @@ inline void assign(Eigen::Matrix<T, 1, Eigen::Dynamic>& x,
+          typename = require_not_same_t<index_uni, I>,
+          typename = require_all_eigen_vector_t<EigVec1, EigVec2>>
+inline void assign(EigVec1& x, const cons_index_list<I, nil_index_list>& idxs,
+                   const EigVec2& y, const char* name = "ANON", int depth = 0) {
   math::check_size_match("vector[multi] assign sizes", "lhs",
                          rvalue_index_size(idxs.head_, x.size()), name,
                          y.size());
+  const Eigen::Ref<const typename EigVec2::PlainObject>& vec = y;
   for (int n = 0; n < y.size(); ++n) {
     int i = rvalue_at(n, idxs.head_);
     math::check_range("vector[multi] assign range", name, x.size(), i);
-    x(i - 1) = y(n);
+    x(i - 1) = vec(n);
   }
 }
 
 /**
- * Assign the specified Eigen row vector at the specified multiple
- * index to the specified value.
+ * Assign the specified Eigen matrix at the specified single index
+ * to the specified row vector value.
  *
- * Types:   row_vec[multi] <- row_vec
+ * Types:  mat[uni,] = rowvec
  *
- * @tparam T Scalar type for assigned row vector.
- * @tparam I Type of multiple index.
- * @tparam U Type of value row vector scalar (must be assignable
- * to T).
- * @param[in] x Row vector variable to be assigned.
- * @param[in] idxs Sequence of one multiple index (from 1).
- * @param[in] y Value vector.
+ * @tparam T Assigned matrix scalar type.
+ * @tparam U Type of value scalar for row vector (must be
+ * assignable to T).
+ * @param[in] x Matrix variable to be assigned.
+ * @param[in] idxs Sequence of one single index (from 1).
+ * @param[in] y Value row vector.
  * @param[in] name Name of variable (default "ANON").
  * @param[in] depth Indexing depth (default 0).
  * @throw std::out_of_range If any of the indices are out of bounds.
- * @throw std::invalid_argument If the value size isn't the same as
- * the indexed size.
+ * @throw std::invalid_argument If the number of columns in the row
+ * vector and matrix do not match.
  */
-template <typename T, typename I, typename U>
-inline typename boost::disable_if<boost::is_same<I, index_uni>, void>::type
-assign(Eigen::Matrix<T, 1, Eigen::Dynamic>& x,
-       const cons_index_list<I, nil_index_list>& idxs,
-       const Eigen::Matrix<U, 1, Eigen::Dynamic>& y, const char* name = "ANON",
-       int depth = 0) {
-  math::check_size_match("row_vector[multi] assign sizes", "lhs",
-                         rvalue_index_size(idxs.head_, x.size()), name,
-                         y.size());
-  for (int n = 0; n < y.size(); ++n) {
-    int i = rvalue_at(n, idxs.head_);
-    math::check_range("row_vector[multi] assign range", name, x.size(), i);
-    x(i - 1) = y(n);
-  }
+template <typename T, typename RowVec,
+          typename = require_t<is_eigen_row_vector<RowVec>>>
+void assign(Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& x,
+            const cons_index_list<index_uni, nil_index_list>& idxs,
+            const RowVec& y, const char* name = "ANON", int depth = 0) {
```

Comment:
@t4c1 should assign here be able to assign to an eigen expression? I left it as Eigen::Matrix since I'm not sure if we ever do something like `assign(A.array() * 2, B.array() + 2)`

Reply:
We don't and this example would not work in Eigen anyway. However, we could let it assign to expressions such as transpose or block. But to use such expressions we would probably need some modifications to compiler and possibly stan language.

In other words I don't think we need it yet, but it may come in handy in the future.